lib/puppetserver/utils/file_utils.rb

```diff
@@ -0,0 +1,64 @@
+require 'fileutils'
+
+module Puppetserver
+  module Utils
+    module FileUtilities
+
+      def self.ensure_dir(setting)
+        if !File.exist?(setting)
+          FileUtils.mkdir_p(setting, mode: 0750)
+          user, group = find_user_and_group
+          FileUtils.chown(user, group, setting)
+        end
+      end
+
+      def self.ensure_file(location, content, mode)
+        if !File.exist?(location)
+          Puppetserver::Utils::FileUtilities.write_file(location, content, mode)
+        end
+      end
+
+      def self.validate_file_paths(one_or_more_paths)
+        return [] unless one_or_more_paths.any?
+        errors = []
+        Array(one_or_more_paths).each do |path|
+          if !File.exist?(path) || !File.readable?(path)
+            errors << "Could not read file '#{path}'"
+          end
+        end
+
+        errors
+      end
+
+      def self.find_user_and_group
+        if !running_as_root?
+          return Process.euid, Process.egid
+        else
+          if pe_puppet_exists?
+            return 'pe-puppet', 'pe-puppet'
+          else
+            return 'puppet', 'puppet'
+          end
+        end
+      end
+
+      def self.running_as_root?
+        !Gem.win_platform? && Process.euid == 0
+      end
+
+      def self.pe_puppet_exists?
+        !!(Etc.getpwnam('pe-puppet') rescue nil)
+      end
+
+      def self.write_file(path, one_or_more_objects, mode)
+        File.open(path, 'w', mode) do |f|
+          Array(one_or_more_objects).each do |object|
+            f.puts object.to_s
+          end
+        end
+        user, group = find_user_and_group
```

Comment:
This is a potentially expensive method† that we call every time someone calls write_file. I think it would make sense to either have the caller call find_user_and_group, save those values, and then pass them into {ensure,write}_file, or make this a utility class that saves the user and group as some state within it.

† Well, pe_puppet_exists? raises and rescues if the pe-puppet user doesn't exist, which doesn't take a lot of time but is a pretty bad practice and I don't know a better way to do it considering Etc's interface